src/ros2idl.ne

```diff
@@ -0,0 +1,395 @@
+  INTEGER: /[0-9]+/,
+  COMMENT: /(?:\/\/[^\n]*)|(?:\/\*(?:.|\n)+?\*\/)/,
+  HEX_LITERAL: /0x(?:[0-9][a-f][A-F])+?/,
+  STRING: [{match: /"(?:\\["\\rnu]|[^"\\])*?"/, lineBreaks: true, value: x => x.slice(1, -1)}], // remove outside quotes
+  LCBR: '{',
+  RCBR: '}',
+  LBR: '[',
+  RBR: ']',
+  LT: '<',
+  GT: '>',
+  LPAR: '(',
+  RPAR: ')',
+  ';': ';',
+  ',': ',',
+  AT: '@',
+  PND: '#',
+  PT: ".",
+  '/': "/",
+  SIGN: /[+|-]/,
+  EQ: /=[^\n]*?/,
+  NAME: {match: /[a-zA-Z_][a-zA-Z0-9_]*(?:\:\:[a-zA-Z][a-zA-Z0-9_]*)*/, type: moo.keywords(kwObject)},
+});
+
+// Utiility functions
+
+// also used to parse tokens to strings since they start as an object
+function join(d){
+	return d.join("");
+}
+
+// used for combining AST components
+function extend(objs) {
+  return objs.reduce((r, p) => ({ ...r, ...p }), {});
+}
+
+function noop() {
+	return null;
+}
+
+// Constants can be used in defs to define lengths of arrays and strings
+// they should be defined before the def that uses them
+const constantToValueMap = {};
+
+function getIntOrConstantValue(d) {
+	const int = parseInt(d);
+	if(!isNaN(int)) {
+		return int
+	}
+
+	// handle %NAME value
+	return d?.value ? constantToValueMap[d.value] : undefined;	
+}
+
+function processComplexModule(d) {
+  const moduleName = d[0][4].name;
+  const defs = d[0][8];
+  // returning array of message definitions
+  const msgDefs = [];
+  function traverse(node, processNode) {
+    if(Array.isArray(node)) {
+      node.forEach(n => traverse(n, processNode));
+    } else {
+      processNode(node);
+    }
+  }
+  // Need to update the names of modules and structs to be in their respective namespaces
+  traverse(d[0], (sub) => {
+    if(sub && sub.definitions) {
+      sub.name = `${moduleName}::${sub.name}`;
+      msgDefs.push(sub);
+    }
+  });
+  
+  
+  return msgDefs;
+}
+
+function processConstantModule(d) {
+	const moduleName = d[0][4].name;
+  const enclosedConstants = d[0][8];
+  // need to return array here to keep same signature as processComplexModule
+	return [{
+		name: moduleName,
+		definitions: enclosedConstants.flat(1),
+	}];
+}
+
+%}
+
+@lexer lexer
+
+main -> (importDcl __):* module:+ _ {% d => {
+	return d[1][0];
+}
+%}
+
+# support <import> or "import" includes - just ignored
+importDcl -> _ "#" "include" __ (%STRING | "<" _ %NAME ("/" %NAME):* "." "idl" _ ">") {% noop %}
+
+# constant modules need to be separate from complex modules since they shouldn't mix
+module  -> ((comment|annotation):* _ "module" __ fieldName __ "{" __ (constantDcl):+ __ "}" semi) {% processConstantModule %}
+  | ((comment|annotation):* _ "module" __ fieldName __ "{" __ (structWithAnnotation|module):+ __ "}" semi) {% processComplexModule %}
+
+
+structWithAnnotation -> (comment|annotation):* struct {% 
+ // ignore annotations on structs because we can only read default value (which doesn't apply)
+ d => {return d[1];}
+%}
+
+struct -> _ "struct" __ fieldName __ "{" __ (declaration):+ __ "}" semi {% d => {
```

Comment:
just guessing that these WS are not actually required?
```suggestion
struct -> _ "struct" __ fieldName _ "{" _ (declaration):+ _ "}" _ semi {% d => {
```

Reply:
actually looks like we might just be able to ignore whitespace completely along with comments, if we have the right tokens. from IDL spec:
>Blanks, horizontal and vertical tabs, newlines, form feeds, and comments (collective, "white space") as described
below are ignored except as they serve to separate tokens. Some white space is required to separate otherwise adjacent identifiers, keywords, and constants.